crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -535,6 +572,63 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_nodes_disruptive(to_upgrade)
+        to_upgrade.each do |element|
+          # If role has a cluster assigned ugprade that cluster (reuse the non-disruptive code here)
+          if ServiceObject.is_cluster?(element)
+            cluster = ServiceObject.cluster_name element
+            cluster_env = "pacemaker-config-#{cluster}"
+            founder = ::Node.find(
+              "pacemaker_founder:true AND " \
+              "pacemaker_config_environment:#{cluster_env}"
+            ).first
+            upgrade_cluster founder, cluster_env
+          else
+            # if cluster(s) and node(s) assigned to the role are already upgraded, skip to next role
+            # if role has single node(s) assigned upgrade those nodes (optionally all at once)
+            upgrade_one_node element
+            # FIXME: this would be nice to do in paralel...
+          end
+        end
+
+        # FIXME: (maybe) skip cinder-volume nodes here that also have a nova compute role assigned
+        # and handle those as part of the compute node upgrade
+      end
+
+      # Go through active proposals and return elements with assigned roles in the right order
+      def upgradable_elements_of_proposals(proposals)
+        to_upgrade = []
+        # If barclamp has an active proposal go through the roles in element order
+        # (skip the compute roles)
+        proposals.each do |name, propmap|
+          proposal = propmap[:prop]
+          elements = proposal["deployment"][name]["elements"]
+          proposal["deployment"][name]["element_order"].flatten.each do |el|
+            next if el.start_with?("nova-compute") || elements[el].nil?
+            to_upgrade |= elements[el]
+          end
+        end
+        to_upgrade
+      end
+
+      def upgrade_disruptive_controllers
+        save_upgrade_state("Entering disruptive upgrade of controller nodes")
+
+        # Go through all OpenStack barclamps (skip "pacemaker" and "openstack") ordered by run_order
```

Comment:
> Here I'm not sure what was the meaning of "openstack" here, there's no barclamp like that... @rhafer ?

The ```prop.category != "OpenStack"``` takes information from the ```config/catalog.yml```. Which itself is generated from the various ```<barclampname>.yml``` in the git repositores. There you can define, that a certain barclamp is a "member" of another barclamp. ```prop.category != "OpenStack"``` matches all barclamps that are member of the "OpenStack" barclamp (yes, there is a barclamp called openstack, it doesn't do much and does really provide a UI)

> It's true that we are skipping all nodes that do not have any openstack roles assigned here... that was not intentional, but I think it could be solved at the end of upgrade by a check for 'anything that is left'

What would that be?